Upstreamize API guide (#3212)
theforeman/foreman-documentation: guides/doc-Project_API/topics/proc_uploading-content-to-projectserver.adoc

[id="uploading-content-to-{project-context}-server"]
= Uploading Content to {ProjectServer}

> You can use the {Project} API to upload and import large files to your {ProjectServer}.

Reviewer: does this work for any file type? Needs technical ACK pls.

Author: Yes, it does. Verified in APIdoc:
![image](https://github.com/user-attachments/assets/b398266b-0a61-46e3-a716-dd4041e78735)

This process involves four steps:

. Create an upload request.
. Upload the content.
. Import the content.
. Delete the upload request.

The maximum file size that you can upload is 2{nbsp}MB.
For information about uploading larger content, see xref:exam-API_Guide-Uploading_Content_Larger_than_2_MB[].

[id="proc-API_Guide-Uploading_Content_to_the_{Project}_Server"]
.Procedure
. Assign the package name to the variable `name`:
+
Example request:
+
[options="nowrap", subs="+quotes,attributes"]
----
$ export name=jq-1.6-2.el7.x86_64.rpm
----
. Assign the checksum of the file to the variable `checksum`:
+
Example request:
+
[options="nowrap", subs="+quotes,attributes"]
----
$ export checksum=$(sha256sum $name|cut -c 1-65)
----
. Assign the file size to the variable `size`:
+
Example request:
+
[options="nowrap", subs="+quotes,attributes"]
----
$ export size=$(du -bs $name|cut -f 1)
----

> . The following command creates a new upload request and returns the upload ID of the request using `size` and `checksum`.

Reviewer: ```suggestion
. Create an upload request that returns the upload ID of the request using `size` and `checksum`.
```

+
Example request:
+
[options="nowrap", subs="+quotes,attributes"]
----
$ curl -H 'Content-Type: application/json' -X POST -k \
-u _sat_username:sat_password_ \
-d "{\"size\": \"$size\", \"checksum\":\"$checksum\"}" \

> https://$(hostname -f)/katello/api/v2/repositories/76/content_uploads

Reviewer: ```suggestion
https://_{foreman-example-com}_/katello/api/v2/repositories/76/content_uploads
```

same below

Content-Type:multipart/form-data \
-X PUT --data-urlencode size=$size --data-urlencode offset=6000000
--data-urlencode content@$\{path}3 https://$(hostname -f)/katello/api/v2/repositories/76/content_uploads/$upload_id

$ curl -u _sat_username:sat_password_ -H Accept:application/json -H \
Content-Type:multipart/form-data \
-X PUT --data-urlencode size=$size --data-urlencode offset=8000000 \
--data-urlencode content@$\{path}4 https://$(hostname -f)/katello/api/v2/repositories/76/content_uploads/$upload_id
----
. Import the complete upload to the repository:
+
[options="nowrap", subs="+quotes,attributes"]
----
$ curl -H "Content-Type:application/json" -X PUT -u \
_sat_username:sat_password_ -k -d \
"{\"uploads\":[{\"id\": \"$upload_id\", \
\"name\": \"$name\", \"checksum\": \"$checksum\" }]}" \
https://$(hostname -f)/katello/api/v2/repositories/76/import_uploads
----
. Delete the upload request:
+
[options="nowrap", subs="+quotes,attributes"]
----
$ curl -H 'Content-Type: application/json' -X DELETE -k \
-u _sat_username:sat_password_ -d "{}" \
https://$(hostname -f)/katello/api/v2/repositories/76/content_uploads/$upload_id
----

[id="exam-API_Guide-Uploading_Duplicate_Content"]

> .Uploading duplicate content

Reviewer: OK for now. Future improvement: maybe move to a troubleshooting section.
